rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,658 @@
+- Dev commands (like adding items): immediately alter the game world
+
+As a result, we require two distinct (but inter-related) abstractions to handle these patterns.
+
+Regardless of this classification, dev tools are:
+
+- not intended to be shipped to end users
+  - as a result, they can be enabled or disabled at compile time, generally behind a single `dev` flag in the application
+- highly application specific: the tools needed for a 3D FPS, a 2D tilemap game, or a puzzle game are all very different
+  - adding these to your game should be done granularly at compile time
+- enabled or operated one at a time, on an as-needed basis, when problems arise
+  - toggling dev tools needs to be done at runtime: without recompiling or restarting the game
+- intended to avoid interfering with the games ordinary operation, especially when disabled
+  - keybindings must be configurable to avoid conflicting with actual gameplay
+  - while off, performance costs should be low or zero
+- expansive and ad hoc: it's common to end up with dozens of assorted dev tools in a finished game, added to help solve specific problems as they occur
+  - some structure is generally needed to present these options to the developer, and avoid ovewhelming them with dozens of chorded hotkeys
+
+In order to manage the complexity of an eclectic collection of one-off debugging tools, dev tools are commonly organized into a **toolbox**: an abstraction used to provide a unified interface over the available dev tools.
+This might be:
+
+- an in-game dev console (like in Quake)
+- a command palette in an editor
+- special developer commands entered into the chat box, commonly with a `/` to denote their non-textual effect
+- cheat codes
+- a unified set of hotkeys
+- game editor widgets
+
+Regardless of the exact interface used, toolboxes have a few shared needs:
+
+- turn on and off various modes
+- ensure that the active modes don't interfere with each other
+- execute one-time commands (like spawning enemies or setting the player's gold)
+- list the various options for the user, ideally with help text
+
+### Toggling modes: `ModalDevTool`
+
+In order to facilitate the creation of toolboxes, Bevy provides the `ModalDevTool` trait.
+
+```rust
+/// Modal dev tools are used by developers to inspect their application in a toggleable way,
+/// such as an FPS meter or a fly camera.
+/// 
+/// Their configuration is stored as a resource (the type that this trait is implemented for),
+/// and they can be enabled, disabled and reconfigured at runtime.
+/// 
+/// The documentation on this struct is reflected, and can be read by toolboxes to provide help text to users.
+trait ModalDevTool: Resource + Reflect + FromReflect + FromStr<Err=DevToolParseError> + Debug {
+    /// The name of this tool, as might be supplied by a command line interface.
+    fn name() -> &'static str {
+        Self::type_name().to_snake_case()
+    }
```

Comment:
I don't think we can have this trait as object safe at all, we can't have empty methods here, all them must accept &mut self or &self

Let alone those methods, we can't have FromStr and Debug as bounds because those traits aren't object safe, and I'm not sure if Resource is object safe at all, maybe we are forced to reflect all we can and have a FromStr reflected somehow.

Reply:
I think impl will be more easy if we have this as object safe, So we might have the name inside metadata and maybe call &self in metadata so we can be object safe?